File: frontend/src/types/f1.ts
// --- Existing types (retained) ---

export interface DriverComparison {
    driver1: {
        name: string;
        avg_lap: string;
        best_lap: string;
        consistency: string;
        fuel_data?: {
            start_fuel: number;
            end_fuel: number;
            fuel_usage: number;
        } | null;
    };
    driver2: {
        name: string;
        avg_lap: string;
        best_lap: string;
        consistency: string;
        fuel_data?: {
            start_fuel: number;
            end_fuel: number;
            fuel_usage: number;
        } | null;
    };
}

export interface ComparisonFormData {
    year: number;
    grand_prix: string;
    session: string;
    driver1: string;
    driver2: string;
}

export interface YearDataResponse {
    grand_prix: string[];
    sessions: string[];
    drivers: string[];
    driver_names: string[];
}

export interface CalendarEvent {
    round_number: number;
    event_name: string;
    event_format: string;
    meeting_key: string;
}

export interface YearCalendarResponse {
    year: number;
    total_rounds: number;
    calendar: CalendarEvent[];
}

export interface SessionInfo {
    type: string;
    date: string;
    session_key: string;
}

export interface GPSessionsResponse {
    year: number;
    grand_prix: string;
    event_format: string;
    sessions: SessionInfo[];
}

export interface SessionDriversResponse {
    grand_prix: string;
    year: number;
    drivers: string[];
}

// --- Comparison types ---

export interface LapTimeEntry {
    lap_number: number;
    driver: string;
    time_seconds: number;
    compound?: string;
}

export interface ComparisonLapsResponse {
    driver1: string;
    driver2: string;
    laps: LapTimeEntry[];
}

export interface TelemetryPoint {
    distance: number;
    speed: number;
    throttle: number;
    brake: boolean;
}

export interface DriverTelemetry {
    driver: string;
    team_color: string;
    data: TelemetryPoint[];
}

export interface ComparisonTelemetryResponse {
    driver1: DriverTelemetry;
    driver2: DriverTelemetry;
}

export interface SectorTime {
    driver: string;
    sector1: number;
    sector2: number;
    sector3: number;
}

export interface ComparisonSectorsResponse {
    driver1: SectorTime;
    driver2: SectorTime;
}

export interface Stint {
    stint_number: number;
    compound: string;
    lap_start: number;
    lap_end: number;
    laps: number;
}

export interface DriverStrategy {
    driver: string;
    stints: Stint[];
}

export interface ComparisonStrategyResponse {
    driver1: DriverStrategy;
    driver2: DriverStrategy;
}

export interface PositionEntry {
    lap_number: number;
    driver: string;
    position: number;
}

export interface ComparisonPositionsResponse {
    driver1: string;
    driver2: string;
    positions: PositionEntry[];
}

export interface ComparisonQueryParams {
    year: number;
    gp: string;
    session: string;
    driver1: string;
    driver2: string;
}

// --- Standings types ---

export interface DriverStanding {
    position: number;
    driver: string;
    driver_abbreviation: string;
    team: string;
    team_color: string;
    points: number;
}

export interface StandingsResponse {
    year: number;
    standings: DriverStanding[];
}

export interface ConstructorStanding {
    position: number;
    team: string;
    team_color: string;
    points: number;
}

export interface ConstructorStandingsResponse {
    year: number;
    standings: ConstructorStanding[];
}

// --- Dashboard types ---

export interface NextRace {
    event_name: string;
    round_number: number;
    date: string;
    country?: string;
}

export interface LastRaceResult {
    event_name: string;
    top5: DriverStanding[];
}

export interface DashboardOverview {
    year: number;
    next_race?: NextRace;
    last_race?: LastRaceResult;
    driver_standings_top10: DriverStanding[];
    constructor_standings_top10: ConstructorStanding[];
}

// --- Request type for AI analysis ---

export interface CompareDriversRequest {
    year: number;
    grand_prix: string;
    session: string;
    driver1: string;
    driver2: string;
}

File: backend/app/schemas/f1.py
from pydantic import BaseModel, ConfigDict
from typing import List, Optional
from datetime import datetime

class BaseResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        json_schema_extra={
            "example": {
                "message": "Success"
            }
        }
    ) 

class CompareDriversRequest(BaseModel):
    year: int
    grand_prix: str
    session: str
    driver1: str
    driver2: str

    model_config = {
        "json_schema_extra": {
            "example": {
                "year": 2024,
                "grandPrix": "Monaco",
                "session": "FP1",
                "driver1": "VER",
                "driver2": "HAM"
            }
        }
    }

class DriverComparisonResponse(BaseModel):
    analysis: str

class SessionInfo(BaseModel):
    type: str
    date: str

    model_config = {
        "json_schema_extra": {
            "example": {
                "type": "FP1",
                "date": "2024-05-24 10:30:00"
            }
        }
    }

class CalendarEvent(BaseModel):
    round_number: int
    event_name: str
    event_format: str

    model_config = {
        "json_schema_extra": {
            "example": {
                "roundNumber": 1,
                "eventName": "Bahrain Grand Prix",
                "eventFormat": "conventional"
            }
        }
    }

class YearCalendarResponse(BaseModel):
    year: int
    total_rounds: int
    calendar: List[CalendarEvent]

class GPSessionsResponse(BaseModel):
    year: int
    grand_prix: str
    event_format: str
    sessions: List[SessionInfo]

class SessionDriversResponse(BaseModel):
    grand_prix: str
    year: int
    drivers: List[str]

    model_config = {
        "json_schema_extra": {
            "example": {
                "grandPrix": "Monaco",
                "year": 2024,
                "drivers": ["VER", "HAM", "LEC"]
            }
        }
    } 

class YearDataResponse(BaseModel):
    grand_prix: List[str]
    drivers: List[str]
    driver_names: List[str]

class LapTimeEntry(BaseModel):
    lap_number: int
    driver: str
    time_seconds: float
    compound: Optional[str] = None

class ComparisonLapsResponse(BaseModel):
    driver1: str
    driver2: str
    laps: List[LapTimeEntry]

class TelemetryPoint(BaseModel):
    distance: float
    speed: float
    throttle: float
    brake: bool

class DriverTelemetry(BaseModel):
    driver: str
    team_color: str
    data: List[TelemetryPoint]

class ComparisonTelemetryResponse(BaseModel):
    driver1: DriverTelemetry
    driver2: DriverTelemetry

class SectorTime(BaseModel):
    driver: str
    sector1: float
    sector2: float
    sector3: float

class ComparisonSectorsResponse(BaseModel):
    driver1: SectorTime
    driver2: SectorTime

class Stint(BaseModel):
    stint_number: int
    compound: str
    lap_start: int
    lap_end: int
    laps: int

class DriverStrategy(BaseModel):
    driver: str
    stints: List[Stint]

class ComparisonStrategyResponse(BaseModel):
    driver1: DriverStrategy
    driver2: DriverStrategy

class PositionEntry(BaseModel):
    lap_number: int
    driver: str
    position: int

class ComparisonPositionsResponse(BaseModel):
    driver1: str
    driver2: str
    positions: List[PositionEntry]

class ComparisonQueryParams(BaseModel):
    year: int
    gp: str
    session: str
    driver1: str
    driver2: str

class DriverStanding(BaseModel):
    position: int
    driver: str
    driver_abbreviation: str
    team: str
    team_color: str
    points: float

class StandingsResponse(BaseModel):
    year: int
    standings: List[DriverStanding]

class ConstructorStanding(BaseModel):
    position: int
    team: str
    team_color: str
    points: float

class ConstructorStandingsResponse(BaseModel):
    year: int
    standings: List[ConstructorStanding]

class NextRace(BaseModel):
    event_name: str
    round_number: int
    date: str
    country: Optional[str] = None

class LastRaceResult(BaseModel):
    event_name: str
    top5: List[DriverStanding]

class DashboardOverview(BaseModel):
    year: int
    next_race: Optional[NextRace] = None
    last_race: Optional[LastRaceResult] = None
    driver_standings_top10: List[DriverStanding]
    constructor_standings_top10: List[ConstructorStanding]